File: Pod/Classes/PXLPhoto.m
//
//  PXLPhoto.m
//  Pods
//
//  Created by Tim Shi on 4/30/15.
//
//

#import "PXLPhoto.h"

#import "PXLProduct.h"

@implementation PXLPhoto

+ (instancetype)photoFromDict:(NSDictionary *)dict inAlbum:(PXLAlbum *)album {
    PXLPhoto *photo = [self new];
    photo.identifier = dict[@"id"];
    photo.photoTitle = dict[@"photo_title"];
    if ([dict[@"latitude"] isKindOfClass:[NSNumber class]] && [dict[@"longitude"] isKindOfClass:[NSNumber class]]) {
        CLLocationDegrees latitude = [dict[@"latitude"] doubleValue];
        CLLocationDegrees longitude = [dict[@"longitude"] doubleValue];
        photo.coordinate = CLLocationCoordinate2DMake(latitude, longitude);
    }
    photo.taggedAt = [NSDate dateWithTimeIntervalSince1970:([dict[@"tagged_at"] doubleValue] / 1000)];
    photo.emailAddress = [self valueOrNilFromDict:dict forKey:@"email_address"];
    photo.instagramFollowers = [dict[@"instagram_followers"] integerValue];
    photo.twitterFollowers = [dict[@"twitter_followers"] integerValue];
    photo.avatarUrl = [self nilSafeUrlFromDict:dict forKey:@"avatar_url"];
    photo.username = dict[@"user_name"];
    photo.connectedUserId = [dict[@"connected_user_id"] integerValue];
    photo.source = dict[@"source"];
    photo.contentType = dict[@"content_type"];
    photo.dataFileName = dict[@"data_file_name"];
    photo.mediumUrl = [self nilSafeUrlFromDict:dict forKey:@"medium_url"];
    photo.bigUrl = [self nilSafeUrlFromDict:dict forKey:@"big_url"];
    photo.thumbnailUrl = [self nilSafeUrlFromDict:dict forKey:@"thumbnail_url"];
    photo.sourceUrl = [self nilSafeUrlFromDict:dict forKey:@"source_url"];
    photo.mediaId = dict[@"media_id"];
    photo.existIn = [dict[@"exist_in"] integerValue];
    photo.collectTerm = dict[@"collect_term"];
    photo.albumPhotoId = dict[@"album_photo_id"];
    photo.likeCount = [dict[@"like_count"] integerValue];
    photo.shareCount = [dict[@"share_count"] integerValue];
    photo.actionLink = [self nilSafeUrlFromDict:dict forKey:@"action_link"];
    photo.actionLinkText = [self valueOrNilFromDict:dict forKey:@"action_link_text"];
    photo.actionLinkTitle = [self valueOrNilFromDict:dict forKey:@"action_link_title"];
    photo.actionLinkPhoto = [self valueOrNilFromDict:dict forKey:@"action_link_photo"];
    photo.updatedAt = [NSDate dateWithTimeIntervalSince1970:([dict[@"updated_at"] doubleValue] / 1000)];
    photo.isStarred = [dict[@"is_starred"] boolValue];
    photo.approved = [dict[@"approved"] boolValue];
    photo.archived = [dict[@"archived"] boolValue];
    photo.isFlagged = [dict[@"is_flagged"] boolValue];
    photo.album = album;
    photo.unreadCount = [dict[@"unread_count"] integerValue];
    photo.albumActionLink = [self nilSafeUrlFromDict:dict forKey:@"album_action_link"];
    photo.title = dict[@"title"];
    photo.messaged = [dict[@"messaged"] boolValue];
    photo.hasPermission = [dict[@"has_permission"] boolValue];
    photo.awaitingPermission = [dict[@"awaiting_permission"] boolValue];
    photo.instUserHasLiked = [dict[@"inst_user_has_liked"] boolValue];
    photo.platformLink = [self nilSafeUrlFromDict:dict forKey:@"platform_link"];
    photo.products = [PXLProduct productsFromArray:dict[@"products"] withPhoto:photo];
    return photo;
}

+ (id)valueOrNilFromDict:(NSDictionary *)dict forKey:(NSString *)key {
    id value = dict[key];
    if ([value isKindOfClass:[NSNull class]]) {
        value = nil;
    }
    return value;
}

+ (NSURL *)nilSafeUrlFromDict:(NSDictionary *)dict forKey:(NSString *)key {
    NSString *urlString = dict[@"key"];
    if (urlString) {
        return [NSURL URLWithString:urlString];
    }
    return nil;
}

@end

File: Sources/PixleeSDK/PXLPhoto.swift
//
//  PXLPhoto.swift
//  pixlee_sdk
//
//  Created by Csaba Toth on 2020. 01. 07..
//  Copyright © 2020. BitRaptors. All rights reserved.
//

import Alamofire
import Foundation
import MapKit

public enum PXLPhotoSize {
    case thumbnail
    case medium
    case big
    case original
}

public struct PXLPhoto: Equatable {
    public static func == (lhs: PXLPhoto, rhs: PXLPhoto) -> Bool {
        lhs.id == rhs.id
    }

    public let id: Int
    public let photoTitle: String?
    public let latitude: Double?
    public let longitude: Double?
    public let taggedAt: Date?
    public let emailAddress: String?
    public let instagramFollowers: Int?
    public let twitterFollowers: Int?
    public let avatarUrl: URL?
    public let username: String?
    public let connectedUserId: Int?
    public let source: String?
    public let contentType: String?
    public let dataFileName: String?
    public let mediumUrl: URL?
    public let bigUrl: URL?
    public let thumbnailUrl: URL?
    public let sourceUrl: URL?
    public let mediaId: String?
    public let existIn: Int?
    public let collectTerm: String?
    public let albumPhotoId: Int
    public let albumId: Int
    public let likeCount: Int?
    public let shareCount: Int?
    public let actionLink: URL?
    public let actionLinkText: String?
    public let actionLinkTitle: String?
    public let actionLinkPhoto: String?
    public let updatedAt: Date?
    public let isStarred: Bool?
    public let approved: Bool
    public let archived: Bool?
    public let isFlagged: Bool?
    public let unreadCount: Int?
    public let albumActionLink: URL?
    public let title: String?
    public let messaged: Bool?
    public let hasPermission: Bool?
    public let awaitingPermission: Bool?
    public let instUserHasLiked: Bool?
    public let platformLink: URL?
    public let products: [PXLProduct]?
    public let cdnSmallUrl: URL?
    public let cdnMediumUrl: URL?
    public let cdnLargeUrl: URL?
    public let cdnOriginalUrl: URL?
    public let cdnSquareMediumUrl: URL?
    public let uploaderAdditionalFields: [String: Any]?
    public let boundingBoxProducts: [BoundingBoxProduct]?

    public var coordinate: CLLocationCoordinate2D? {
        guard let latitude = self.latitude, let longitude = self.longitude else { return nil }
        return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    public func photoUrl(for size: PXLPhotoSize) -> URL? {
        if contentType == "video" {
            return getResizedPhotoUrl(for: size)
        }

        if approved {
            return getCDNPhotoUrl(for: size)
        } else {
            return getResizedPhotoUrl(for: size)
        }
    }

    public var isVideo: Bool {
        return contentType == "video"
    }

    public func videoUrl() -> URL? {
        if contentType == "video" {
            return sourceUrl
        }
        return nil
    }

    func getResizedPhotoUrl(for size: PXLPhotoSize) -> URL? {
        switch size {
        case .thumbnail:
            return thumbnailUrl
        case .medium:
            return mediumUrl
        case .big:
            return bigUrl
        case .original:
            return sourceUrl
        }
    }

    func getCDNPhotoUrl(for size: PXLPhotoSize) -> URL? {
        switch size {
        case .thumbnail:
            return cdnSmallUrl
        case .medium:
            return cdnMediumUrl
        case .big:
            return cdnLargeUrl
        case .original:
            return cdnOriginalUrl
        }
    }

    public var sourceIconImage: UIImage? {
        // test it in Cocoapods
        #if SWIFT_PACKAGE
        let bundle = Bundle.module
        #else
        let bundle = Bundle(for: PXLAlbum.self)
        #endif
        switch source {
        case "instagram":
            return UIImage(named: "instagram", in: bundle, compatibleWith: nil)
        case "facebook":
            return UIImage(named: "facebook", in: bundle, compatibleWith: nil)
        case "pinterest":
            return UIImage(named: "pinterest", in: bundle, compatibleWith: nil)
        case "tumblr":
            return UIImage(named: "tumblr", in: bundle, compatibleWith: nil)
        case "twitter":
            return UIImage(named: "twitter", in: bundle, compatibleWith: nil)
        case "vine":
            return UIImage(named: "vine", in: bundle, compatibleWith: nil)
        default:
            return nil
        }
    }

    public func triggerEventActionClicked(actionLink: String, completionHandler: @escaping (Error?) -> Void) -> DataRequest {
        return PXLAnalyticsService.sharedAnalytics.logEvent(event: PXLAnalyticsEventActionClicked(photo: self, actionLink: actionLink), completionHandler: completionHandler)
    }

    public func triggerEventOpenedLightbox(completionHandler: @escaping (Error?) -> Void) -> DataRequest {
        return PXLAnalyticsService.sharedAnalytics.logEvent(event: PXLAnalyticsEventOpenedLightBox(photo: self), completionHandler: completionHandler)
    }
}
